File: src/main/java/gameover/fwk/gdx/ai/Point.java
package gameover.fwk.gdx.ai;

/**
 * Class to hold a Point during A* computation.
 */
public class Point {
    int x;
    int y;
    int g;
    int h;
    Point ancestor;
    int index;

    Point(int x, int y, int g, int h, Point ancestor) {
        this.x = x;
        this.y = y;
        this.g = g;
        this.h = h;
        this.ancestor = ancestor;
        this.index = ancestor!=null?ancestor.index+1:0;
    }

    @Override
    public boolean equals(Object obj) {
        Point p = (Point)obj;
        return is(p.x, p.y);
    }

    @Override
    public int hashCode() {
        int result = x;
        result = 31 * result + y;
        return result;
    }

    @Override
    public String toString(){
        if(ancestor==null)
            return x+","+y;
        else
            return x+","+y+"<="+ancestor.toString();
    }

    int f(){
        return g + h;
    }

    public boolean is(int tx, int ty) {
        return tx==x && ty==y;
    }
}

File: src/main/scala/gameover/fwk/ai/Point.scala
package gameover.fwk.ai

/**
 * Class to hold a Point during A* computation.
 */
class Point(val x: Int, val y: Int, var g: Int, var h: Int, var ancestor: Option[Point] = None) {

  val index: Int = ancestor match {
    case Some(p) => p.index + 1
    case None => 0
  }

  def canEqual(a: Any) = a.isInstanceOf[Point]

  override def equals(that: Any): Boolean =
    that match {
      case that: Point => that.canEqual(this) && this.x == that.x && this.y == that.y
      case _ => false
    }


  override def hashCode: Int = {
    41 * x + y
  }

  override def toString: String = {
    ancestor match {
      case Some(p) => s"$x,$y<=$p"
      case None => s"$x,$y"
    }
  }

  def f = g + h

  def is(tx: Int, ty: Int) = tx == x && ty == y
}
